### PR TITLE
Refactor platform detection

### PR DESCRIPTION
**Pull request by umangachapagain**
This adds ability to detect whether the application is running on
OpenShift or not. If it is running on OpenShift, it also detects the
OpenShift platform type.

This also reduces API calls to detect platform every reconcile.
Instead, platform is not initialized once and re-used throughout the
lifecycle of the operator.

### CODE REVIEW

File: controllers/storagecluster/main_test.go [hunk 1]
```diff
@@ -0,0 +1,15 @@
+package storagecluster
+
+import (
+	"os"
+	"testing"
+
+	"github.com/red-hat-storage/ocs-operator/v4/controllers/platform"
+)
+
+func TestMain(m *testing.M) {
```
Comment (iamniting) [inline controllers/storagecluster/main_test.go:10]: Why do we need this?

**Review by iamniting — CHANGES_REQUESTED** [review]

File: controllers/platform/platform_detection.go [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+// platform is used to get the PlatformType of the running cluster in a thread-safe manner
+// It is a singleton which is initialized exactly once via Detect() function call.
+type platform struct {
+	isOpenShift bool
+	platform    configv1.PlatformType
+}
+
+// SetFakePlatformInstanceForTesting can be used to fake a Platform while testing.
+// It should only be used for testing. This is not thread-safe.
+func SetFakePlatformInstanceForTesting(isOpenShift bool, platformType configv1.PlatformType) {
```
[the remark this answers is not in the document]
Comment (umangachapagain) [reply controllers/platform/platform_detection.go:34]: If we do that, other packages won't be able to import it for testing.

It seems defining test functions with other functions is the recommended approach here. We just need to be careful that we do not use it in non-test code.

File: controllers/platform/platform_detection.go [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+	"k8s.io/client-go/discovery"
+	"k8s.io/client-go/rest"
+	ctrl "sigs.k8s.io/controller-runtime"
+)
+
+var (
+	once             sync.Once
+	platformInstance *platform
+
+	ErrorPlatformNotDetected = errors.New("platform not detected")
```
[the remark this answers is not in the document]
Comment (umangachapagain) [reply controllers/platform/platform_detection.go:22]: Errors can not be constants. We get an error for such assignment.

File: controllers/platform/platform_detection.go [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+
+// DevicesDefaultToFastForThisPlatform determines whether we should TuneFastDeviceClass for this platform.
+// It returns false if we shouldn't TuneFastDeviceClass or if platform is not yet detected.
+func DevicesDefaultToFastForThisPlatform() (bool, error) {
+	// tuneFastPlatforms is a list of all PlatformTypes where TuneFastDeviceClass has to be set True.
+	var tuneFastPlatforms = []configv1.PlatformType{
+		configv1.OvirtPlatformType,
+		configv1.IBMCloudPlatformType,
+		configv1.AzurePlatformType,
+	}
```
[the remark this answers is not in the document]
Comment (umangachapagain) [reply controllers/platform/platform_detection.go:123]: We do not need to export it as I do not see the use-case.
The only use case that we have is addressed by `DevicesDefaultToFastForThisPlatform()` utility function.
If we find a use-case where we need "a list of platforms to tune" for other purposes, we can export it at that point.

File: controllers/storagecluster/main_test.go (hunk 1 above)
Comment (umangachapagain) [reply controllers/storagecluster/main_test.go:10]: This is used to set the platform before tests in this test suite are executed.
This helps tests avoid setting platform even when they do not use it. When tests need to set a platform, then they will use test helper functions to do so.

**Review by iamniting — APPROVED** [review]

**Pull request merged**